src/components/sticker.rs

```diff
@@ -0,0 +1,191 @@
+
+        spawn(clone!(@weak self as obj, @weak session => async move {
+            if sticker.sticker.local.is_downloading_completed {
+                obj.load_sticker(&sticker.sticker.local.path, looped, format).await;
+            } else {
+            let file_id = sticker.sticker.id;
+                obj.download_sticker(file_id, &session, looped, format).await
+            }
+        }));
+    }
+
+    pub(crate) fn set_max_size(&self, size: i32) {
+        self.imp().max_size.set(size);
+    }
+
+    pub fn play_animation(&self) {
+        if let Some(animation) = &*self.imp().child.borrow() {
+            if let Some(animation) = animation.downcast_ref::<rlt::Animation>() {
+                if !animation.is_playing() {
+                    // TODO: animated emoji needs to play
+                    // effect when someone clicks on it
+                    animation.play();
+                }
+            }
+        }
+    }
+
+    async fn download_sticker(
+        &self,
+        file_id: i32,
+        session: &Session,
+        looped: bool,
+        format: StickerFormat,
+    ) {
+        match session.download_file(file_id).await {
+            Ok(file) => {
+                self.load_sticker(&file.local.path, looped, format).await;
+            }
+            Err(e) => {
+                log::warn!("Failed to download a sticker: {e:?}");
+            }
+        }
+    }
+
+    async fn load_sticker(&self, path: &str, looped: bool, format: StickerFormat) {
+        let path = path.to_owned();
+        let message_id = self.imp().message_id.get();
+
+        let widget: gtk::Widget = match format {
+            StickerFormat::Tgs => {
+                let animation = rlt::Animation::from_filename(&path);
+                animation.set_loop(looped);
+                animation.use_cache(looped);
+                animation.play();
+                animation.upcast()
+            }
+            StickerFormat::Webp => {
+                let result = gio::spawn_blocking(move || decode_image_from_path(&path))
+                    .await
+                    .unwrap();
+
+                match result {
+                    Ok(texture) => {
+                        let picture = gtk::Picture::new();
+                        picture.set_paintable(Some(&texture));
+                        picture.upcast()
+                    }
+                    Err(e) => {
+                        log::warn!("Error decoding a sticker: {e:?}");
+                        return;
+                    }
+                }
+            }
+            _ => unimplemented!(),
+        };
+
+        if self.imp().message_id.get() != message_id {
+            return;
+        }
+
+        self.set_child(Some(widget));
+    }
+
+    fn set_child(&self, child: Option<gtk::Widget>) {
+        let imp = self.imp();
+
+        if let Some(ref child) = child {
+            child.set_parent(self);
+        }
+
+        imp.child.replace(child);
+    }
```

Comment:
Note that this function doesn't do anything to the previous child, so what this function does is just add another child to this widget. You should unparent the previous one if you don't want it anymore as the widget child.